dl1_data_handler/writer.py

```diff
@@ -742,3 +751,63 @@ def _process_data(self, file_list, output_filename):
                     # Reset event count and increment file count
                     event_count = 0
                     output_file_count += 1
+
+
+def gain_selection(waveform, image, peakpos, cam_id, threshold):
+    """
+    Based on the waveform and threshold, select the proper gain for each pixel.
+    By default, the channel 0 is kept.
+    If a pixel is saturated (value > threshold in the waveform), the channel 1 is used.
+
+    Parameters
+    ----------
+    waveform: array of waveforms of the events
+    image: array of calibrated pixel charges
+    peakpos: array of pixel peak positions
+    cam_id: str
+    threshold: int threshold to change form high gain to low gain
+
+    Returns
+    -------
+    combined_image, combined_peakpos: `(numpy.array, numpy.array)`
+        combined_image.shape = image.shape[1]
+    """
+
+    assert image.shape[0] == 2
+
+    gainsel = ThresholdGainSelector(select_by_sample=True)
+    gainsel.thresholds[cam_id] = threshold
+
+    waveform, gain_mask = gainsel.select_gains(cam_id, waveform)
+    signal_mask = gain_mask.max(axis=1)
+
+    combined_image = image[0].copy()
```

Comment:
I think it might be worth doing this in-place instead of copying, as I don't think we need the original image or peak_pos after we have the combined one. It should save some time. I will post some profiling results in the PR discussion.

Reply:
After profiling this seems unnecessary.